platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/internal/cc/impl/isolated/IsolatedProjectsCompositeBuildIntegrationTest.groovy

```diff
@@ -50,40 +48,98 @@ class IsolatedProjectsCompositeBuildIntegrationTest extends AbstractIsolatedProj
+        isolatedProjectsFails "help"
+
+        then:
+        failureCauseContains("A cycle has been detected in the definition of plugin builds: :root -> :library -> :root.")
+    }
+
+    def "'root(plugin) -> plugins -> root(plugin)' is prohibited because of a plugin build cycle"() {
+        given:
+        def rootBuildDir = file("root")
+
+        includePluginBuild(rootBuildDir.file("settings.gradle"), "../plugins-a")
+        applyPlugins(rootBuildDir.file("build.gradle"), "groovy-gradle-plugin", "plugins-a")
+        file("root/src/main/groovy/foo.gradle") << ""
+
+        includedBuild("plugins-a") {
+            includePluginBuild(settingsScript, "../root")
+            applyPlugins(buildScript, "groovy-gradle-plugin", "foo")
+            srcMainGroovy.file("plugin-a.gradle") << ""
+        }
+
+        when:
+        executer.inDirectory(rootBuildDir) // to have a stable root build name
+        isolatedProjectsFails "help"
+
+        then:
+        failureCauseContains("A cycle has been detected in the definition of plugin builds: :plugins-a -> :root -> :plugins-a.")
+    }
+
+    def "'root -> plugins-a -> plugins-b -> plugins-c -> plugins-a' is prohibited because of a plugin build cycle"() {
+        given:
+        includePluginBuild(settingsFile, "plugins-a")
+        applyPlugins(buildFile, "plugin-a")
 
         includedBuild("plugins-a") {
             includePluginBuild(settingsScript, "../plugins-b")
-            applyPlugins(buildScript, ["groovy-gradle-plugin", "plugin-b"])
+            applyPlugins(buildScript, "groovy-gradle-plugin", "plugin-b")
             srcMainGroovy.file("plugin-a.gradle") << ""
         }
 
         includedBuild("plugins-b") {
             includePluginBuild(settingsScript, "../plugins-c")
-            applyPlugins(buildScript, ["groovy-gradle-plugin", "plugin-c"])
+            applyPlugins(buildScript, "groovy-gradle-plugin", "plugin-c")
             srcMainGroovy.file("plugin-b.gradle") << ""
         }
 
         includedBuild("plugins-c") {
             includePluginBuild(settingsScript, "../plugins-a")
-            applyPlugins(buildScript, ["groovy-gradle-plugin", "plugin-a"])
+            applyPlugins(buildScript, "groovy-gradle-plugin", "plugin-a")
             srcMainGroovy.file("plugin-c.gradle") << ""
         }
 
         when:
-        isolatedProjectsFails("help")
+        isolatedProjectsFails "help"
 
         then:
         failureDescriptionContains("A cycle has been detected in the definition of plugin builds: :plugins-a -> :plugins-b -> :plugins-c -> :plugins-a.")
     }
 
-    def "transitive cycles(start is a plugin) for plugin builds are prohibited"() {
+    def "'root -> plugins-a -> library-b -> library-c -> plugins-a' is prohibited because of a plugin build cycle"() {
         given:
         includePluginBuild(settingsFile, "plugins-a")
-        applyPlugins(buildFile, ["plugin-a"])
+        applyPlugins(buildFile, "plugin-a")
 
         includedBuild("plugins-a") {
```

Comment:
🤔  This plugins build does not apply `groovy-gradle-plugin`, but it does not seem to change the outcome of the test. I wonder if we can simplify all tests then by omitting the `srcMainGroovy` lines, because they are never considered.